From ed4e52b72fd595091fcb2b070a15dee4e50362b7 Mon Sep 17 00:00:00 2001
From: Tsuyoshi Ozawa <ozawa@apache.org>
Date: Mon, 5 Oct 2015 08:15:06 +0900
Subject: [PATCH 1072/2748] HADOOP-11098. [JDK8] Max Non Heap Memory default
 changed between JDK7 and 8.

(cherry picked from commit 30e2f836a26490a24c7ddea754dd19f95b24bbc8)
(cherry picked from commit e6514c6a3b5287c08585aa9cc3ce893c1cc777a8)

Change-Id: I7d44caeff4bd45b195c47897f17e70d94ab2a293
---
 .../apache/hadoop/metrics2/source/JvmMetrics.java  |   15 +++++++++++++--
 .../src/main/webapps/hdfs/dfshealth.html           |    4 ++--
 2 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/source/JvmMetrics.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/source/JvmMetrics.java
index c62caf3..164aa93 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/source/JvmMetrics.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/source/JvmMetrics.java
@@ -60,6 +60,7 @@ synchronized JvmMetrics init(String processName, String sessionId) {
   }
 
   static final float M = 1024*1024;
+  static public final float MEMORY_MAX_UNLIMITED_MB = -1;
 
   final MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
   final List<GarbageCollectorMXBean> gcBeans =
@@ -106,13 +107,23 @@ private void getMemoryUsage(MetricsRecordBuilder rb) {
     Runtime runtime = Runtime.getRuntime();
     rb.addGauge(MemNonHeapUsedM, memNonHeap.getUsed() / M)
       .addGauge(MemNonHeapCommittedM, memNonHeap.getCommitted() / M)
-      .addGauge(MemNonHeapMaxM, memNonHeap.getMax() / M)
+      .addGauge(MemNonHeapMaxM, calculateMaxMemoryUsage(memNonHeap))
       .addGauge(MemHeapUsedM, memHeap.getUsed() / M)
       .addGauge(MemHeapCommittedM, memHeap.getCommitted() / M)
-      .addGauge(MemHeapMaxM, memHeap.getMax() / M)
+      .addGauge(MemHeapMaxM, calculateMaxMemoryUsage(memHeap))
       .addGauge(MemMaxM, runtime.maxMemory() / M);
   }
 
+  private float calculateMaxMemoryUsage(MemoryUsage memHeap) {
+    long max =  memHeap.getMax() ;
+
+     if (max == -1) {
+       return MEMORY_MAX_UNLIMITED_MB;
+     }
+
+    return max / M;
+  }
+
   private void getGcUsage(MetricsRecordBuilder rb) {
     long count = 0;
     long timeMillis = 0;
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/webapps/hdfs/dfshealth.html b/hadoop-hdfs-project/hadoop-hdfs/src/main/webapps/hdfs/dfshealth.html
index 27c4b98..90a52d9 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/webapps/hdfs/dfshealth.html
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/webapps/hdfs/dfshealth.html
@@ -151,11 +151,11 @@
   {/fs}
 </p>
 {#mem.HeapMemoryUsage}
-<p>Heap Memory used {used|fmt_bytes} of {committed|fmt_bytes} Heap Memory. Max Heap Memory is {max|fmt_bytes}. </p>
+<p>Heap Memory used {used|fmt_bytes} of {committed|fmt_bytes} Heap Memory. Max Heap Memory is {@eq key=max value="-1" type="number"}&ltunbonded&gt{:else}{max|fmt_bytes}{/eq}.</p>
 {/mem.HeapMemoryUsage}
 
 {#mem.NonHeapMemoryUsage}
-<p>Non Heap Memory used {used|fmt_bytes} of {committed|fmt_bytes} Commited Non Heap Memory. Max Non Heap Memory is {max|fmt_bytes}. </p>
+<p>Non Heap Memory used {used|fmt_bytes} of {committed|fmt_bytes} Commited Non Heap Memory. Max Non Heap Memory is {@eq key=max value="-1" type="number"}&ltunbonded&gt{:else}{max|fmt_bytes}{/eq}.</p>
 {/mem.NonHeapMemoryUsage}
 
 {#nn}
-- 
1.7.9.5

